### ISSUE TITLE
Permission error running DAG

### ISSUE THREAD

**Issue by hududed** [iusztinpaul/energy-forecasting#10]
Running the `ml_pipeline` DAG on airflow failed at the `run_feature_pipeline` stage.

```
[2023-07-10, 20:04:07 UTC] {process_utils.py:189} INFO -   Created wheel for pyhopshive: filename=PyHopsHive-0.6.4.1.dev0-py3-none-any.whl size=48570 sha256=cc58cad2e6520348f8a554bfec9d584a5de309dfce6d8e656801672923f0ed2d
[2023-07-10, 20:04:07 UTC] {process_utils.py:189} INFO -   Stored in directory: /home/***/.cache/pip/wheels/36/91/cc/cb4be3581fd6cd95b73d2b229d66c8975c2d0a1087fa8703b2
[2023-07-10, 20:04:07 UTC] {process_utils.py:189} INFO - Successfully built fire hopsworks avro hsml hsfs twofish thrift future pyhopshive
[2023-07-10, 20:04:08 UTC] {process_utils.py:189} INFO - Installing collected packages: wcwidth, twofish, pytz, pyhumps, pure-eval, ptyprocess, pickleshare, javaobj-py3, fastjsonschema, executing, dataclasses, confluent-kafka, backcall, zipp, urllib3, tzlocal, typing-extensions, traitlets, tqdm, toolz, termcolor, six, ruamel.yaml.clib, rpds-py, python-dotenv, pyparsing, PyMySQL, pygments, pycryptodomex, pycparser, pyasn1, prompt-toolkit, platformdirs, pexpect, parso, packaging, numpy, multidict, mock, mistune, markupsafe, jsonpointer, jmespath, idna, greenlet, future, fastavro, entrypoints, decorator, colorama, Click, charset-normalizer, certifi, avro, attrs, yarl, thrift, sqlalchemy, scipy, ruamel.yaml, requests, referencing, python-dateutil, pyasn1-modules, pyarrow, orderedmultidict, matplotlib-inline, jupyter-core, jsonpatch, jinja2, jedi, importlib-metadata, fire, cffi, asttokens, stack-data, pyjks, pandas, jsonschema-specifications, furl, cryptography, botocore, s3transfer, pyhopshive, jsonschema, Ipython, nbformat, boto3, altair, hsml, great_expectations, hsfs, hopsworks, feature_pipeline
[2023-07-10, 20:04:31 UTC] {process_utils.py:189} INFO - Successfully installed Click-8.1.4 Ipython-8.14.0 PyMySQL-1.1.0 altair-4.2.2 asttokens-2.2.1 attrs-23.1.0 avro-1.11.0 backcall-0.2.0 boto3-1.28.1 botocore-1.31.1 certifi-2023.5.7 cffi-1.15.1 charset-normalizer-3.2.0 colorama-0.4.6 confluent-kafka-1.9.0 cryptography-41.0.1 dataclasses-0.6 decorator-5.1.1 entrypoints-0.4 executing-1.2.0 fastavro-1.7.3 fastjsonschema-2.17.1 feature_pipeline-0.1.0 fire-0.5.0 furl-2.1.3 future-0.18.3 great_expectations-0.14.13 greenlet-2.0.2 hopsworks-3.2.0 hsfs-3.2.0 hsml-3.2.0 idna-3.4 importlib-metadata-6.8.0 javaobj-py3-0.4.3 jedi-0.18.2 jinja2-3.0.3 jmespath-1.0.1 jsonpatch-1.33 jsonpointer-2.4 jsonschema-4.18.0 jsonschema-specifications-2023.6.1 jupyter-core-5.3.1 markupsafe-2.0.1 matplotlib-inline-0.1.6 mistune-3.0.1 mock-5.0.2 multidict-6.0.4 nbformat-5.9.1 numpy-1.25.1 orderedmultidict-1.0.1 packaging-23.1 pandas-1.5.3 parso-0.8.3 pexpect-4.8.0 pickleshare-0.7.5 platformdirs-3.8.1 prompt-toolkit-3.0.39 ptyprocess-0.7.0 pure-eval-0.2.2 pyarrow-12.0.1 pyasn1-0.5.0 pyasn1-modules-0.3.0 pycparser-2.21 pycryptodomex-3.18.0 pygments-2.15.1 pyhopshive-0.6.4.1.dev0 pyhumps-1.6.1 pyjks-20.0.0 pyparsing-2.4.7 python-dateutil-2.8.2 python-dotenv-1.0.0 pytz-2023.3 referencing-0.29.1 requests-2.31.0 rpds-py-0.8.10 ruamel.yaml-0.17.17 ruamel.yaml.clib-0.2.7 s3transfer-0.6.1 scipy-1.11.1 six-1.16.0 sqlalchemy-2.0.18 stack-data-0.6.2 termcolor-2.3.0 thrift-0.16.0 toolz-0.12.0 tqdm-4.65.0 traitlets-5.9.0 twofish-0.3.0 typing-extensions-4.7.1 tzlocal-5.0.1 urllib3-1.26.16 wcwidth-0.2.6 yarl-1.9.2 zipp-3.16.0
[2023-07-10, 20:04:31 UTC] {process_utils.py:189} INFO - 
[2023-07-10, 20:04:31 UTC] {process_utils.py:189} INFO - [notice] A new release of pip is available: 23.1 -> 23.1.2
[2023-07-10, 20:04:31 UTC] {process_utils.py:189} INFO - [notice] To update, run: /tmp/venvx9zfic8n/bin/python -m pip install --upgrade pip
[2023-07-10, 20:04:31 UTC] {process_utils.py:181} INFO - Executing cmd: /tmp/venvx9zfic8n/bin/python /tmp/venvx9zfic8n/script.py /tmp/venvx9zfic8n/script.in /tmp/venvx9zfic8n/script.out /tmp/venvx9zfic8n/string_args.txt
[2023-07-10, 20:04:31 UTC] {process_utils.py:185} INFO - Output:
[2023-07-10, 20:04:32 UTC] {process_utils.py:189} INFO - Traceback (most recent call last):
[2023-07-10, 20:04:32 UTC] {process_utils.py:189} INFO -   File "/usr/lib/python3.9/pathlib.py", line 1312, in mkdir
[2023-07-10, 20:04:32 UTC] {process_utils.py:189} INFO -     self._accessor.mkdir(self, mode)
[2023-07-10, 20:04:32 UTC] {process_utils.py:189} INFO - FileNotFoundError: [Errno 2] No such file or directory: '/home/hud/projects/energy-forecasting/output'
[2023-07-10, 20:04:32 UTC] {process_utils.py:189} INFO - 
[2023-07-10, 20:04:32 UTC] {process_utils.py:189} INFO - During handling of the above exception, another exception occurred:
[2023-07-10, 20:04:32 UTC] {process_utils.py:189} INFO - 
[2023-07-10, 20:04:32 UTC] {process_utils.py:189} INFO - Traceback (most recent call last):
[2023-07-10, 20:04:32 UTC] {process_utils.py:189} INFO -   File "/usr/lib/python3.9/pathlib.py", line 1312, in mkdir
[2023-07-10, 20:04:32 UTC] {process_utils.py:189} INFO -     self._accessor.mkdir(self, mode)
[2023-07-10, 20:04:32 UTC] {process_utils.py:189} INFO - FileNotFoundError: [Errno 2] No such file or directory: '/home/hud/projects/energy-forecasting'
[2023-07-10, 20:04:32 UTC] {process_utils.py:189} INFO - 
[2023-07-10, 20:04:32 UTC] {process_utils.py:189} INFO - During handling of the above exception, another exception occurred:
[2023-07-10, 20:04:32 UTC] {process_utils.py:189} INFO - 
[2023-07-10, 20:04:32 UTC] {process_utils.py:189} INFO - Traceback (most recent call last):
[2023-07-10, 20:04:32 UTC] {process_utils.py:189} INFO -   File "/usr/lib/python3.9/pathlib.py", line 1312, in mkdir
[2023-07-10, 20:04:32 UTC] {process_utils.py:189} INFO -     self._accessor.mkdir(self, mode)
[2023-07-10, 20:04:32 UTC] {process_utils.py:189} INFO - FileNotFoundError: [Errno 2] No such file or directory: '/home/hud/projects'
[2023-07-10, 20:04:32 UTC] {process_utils.py:189} INFO - 
[2023-07-10, 20:04:32 UTC] {process_utils.py:189} INFO - During handling of the above exception, another exception occurred:
[2023-07-10, 20:04:32 UTC] {process_utils.py:189} INFO - 
[2023-07-10, 20:04:32 UTC] {process_utils.py:189} INFO - Traceback (most recent call last):
[2023-07-10, 20:04:32 UTC] {process_utils.py:189} INFO -   File "/tmp/venvx9zfic8n/script.py", line 89, in <module>
[2023-07-10, 20:04:32 UTC] {process_utils.py:189} INFO -     res = run_feature_pipeline(*arg_dict["args"], **arg_dict["kwargs"])
[2023-07-10, 20:04:32 UTC] {process_utils.py:189} INFO -   File "/tmp/venvx9zfic8n/script.py", line 59, in run_feature_pipeline
[2023-07-10, 20:04:32 UTC] {process_utils.py:189} INFO -     from feature_pipeline import utils, pipeline
[2023-07-10, 20:04:32 UTC] {process_utils.py:189} INFO -   File "/tmp/venvx9zfic8n/lib/python3.9/site-packages/feature_pipeline/utils.py", line 5, in <module>
[2023-07-10, 20:04:32 UTC] {process_utils.py:189} INFO -     from feature_pipeline import settings
[2023-07-10, 20:04:32 UTC] {process_utils.py:189} INFO -   File "/tmp/venvx9zfic8n/lib/python3.9/site-packages/feature_pipeline/settings.py", line 44, in <module>
[2023-07-10, 20:04:32 UTC] {process_utils.py:189} INFO -     OUTPUT_DIR.mkdir(parents=True, exist_ok=True)
[2023-07-10, 20:04:32 UTC] {process_utils.py:189} INFO -   File "/usr/lib/python3.9/pathlib.py", line 1316, in mkdir
[2023-07-10, 20:04:32 UTC] {process_utils.py:189} INFO -     self.parent.mkdir(parents=True, exist_ok=True)
[2023-07-10, 20:04:32 UTC] {process_utils.py:189} INFO -   File "/usr/lib/python3.9/pathlib.py", line 1316, in mkdir
[2023-07-10, 20:04:32 UTC] {process_utils.py:189} INFO -     self.parent.mkdir(parents=True, exist_ok=True)
[2023-07-10, 20:04:32 UTC] {process_utils.py:189} INFO -   File "/usr/lib/python3.9/pathlib.py", line 1316, in mkdir
[2023-07-10, 20:04:32 UTC] {process_utils.py:189} INFO -     self.parent.mkdir(parents=True, exist_ok=True)
[2023-07-10, 20:04:32 UTC] {process_utils.py:189} INFO -   File "/usr/lib/python3.9/pathlib.py", line 1312, in mkdir
[2023-07-10, 20:04:32 UTC] {process_utils.py:189} INFO -     self._accessor.mkdir(self, mode)
[2023-07-10, 20:04:32 UTC] {process_utils.py:189} INFO - PermissionError: [Errno 13] Permission denied: '/home/hud'
```

So far from what I've read `sudo chmod 777 ./logs ./plugins` should solve any permission problems, but it didnt work here - any ideas? `/home/hud/projects/energy-forecasting` is the root folder.

**Comment by iusztinpaul** [comment]
In what folder have you created the logs directory? Is it under /airflow? 

I see the error throwing an error with your local path, not the one from the container, which is weird.

**Comment by hududed** [comment]
Yes I followed:
```
cd airflow
mkdir -p ./logs ./plugins
sudo chmod 777 ./logs ./plugins
echo -e "AIRFLOW_UID=$(id -u)" > .env
echo "ML_PIPELINE_ROOT_DIR=/opt/airflow/dags" >> .env
```

This is in `./dags/.env`:
```
FS_API_KEY = "V3FgR972B..."
FS_PROJECT_NAME = "energy_consumption"
WANDB_API_KEY = "972625..."
WANDB_ENTITY = "mlops"
WANDB_PROJECT = "energy_consumption"
GOOGLE_CLOUD_PROJECT = "energy_consumption"
GOOGLE_CLOUD_BUCKET_NAME = "hourly-batch-predictions"
GOOGLE_CLOUD_SERVICE_ACCOUNT_JSON_PATH = "/opt/airflow/dags/credentials/gcp/energy_consumption/admin-buckets.json"
```

**Comment by iusztinpaul** [comment]
It looks right.

Maybe you have permission issues because of Docker, not the Airflow setup. 

It is hard for me to take a conclusion without more info, but try to look up Docker permission issues when it comes to creating new directories inside a Docker volume.

**Comment by iusztinpaul** [comment]
After a better look, there is something wrong with your paths. In your Docker container, you shouldn't use paths that use your host PC paths, such as: `/home/hud/projects/energy-forecasting/output`.

Instead, you should see paths that use the container's internal file structure, such as: `/opt/airflow/dags/output`.

Probably, that is why you get a permission error. How did you run the Airflow Docker containers?

**Comment by hududed** [comment]
```
docker compose up airflow-init
docker compose --env-file .env up --build -d
```

It's odd since afaik the path is determined in that .env file:
```
AIRFLOW_UID=1000
ML_PIPELINE_ROOT_DIR=/opt/airflow/dags
```

Is this set externally in docker? how do I go about troubleshooting that?

**Comment by iusztinpaul** [comment]
What is happening is extremely weird. It looks like you did everything fine. 

`ML_PIPELINE_ROOT_DIR` is set through `docker compose`, and you can access it in the `airflow worker container`. You should connect to the docker container and check it out:

Use the following command to connect to the docker container:
```shell
docker exec -it <airflow_worker_docker_container_id> sh
```

But FIRST of all, I would delete all the images & containers and recreate everything from scratch. Note that you don't have to delete the `.env` files or external projects. I am talking just about the Airflow containers and images.

**Comment by hududed** [comment]
I retried all of this fresh on another OS and it finally works! Though I did the exactly the same steps so it's hard for me to say where it went wrong. Anyway, I am closing this for now - thanks you've been helpful.

**Comment by iusztinpaul** [comment]
It is bizarre what happened, but I am happy that it finally worked 🔥

**Comment by hududed** [comment]
Ok I knew this was going to be a problem if left unresolved - For some reason I am getting the same errors again. But I did notice this at the beginning of the log for `run_feature_pipeline` task:

```
9156d488da30
*** Found local files:
***   * /opt/airflow/logs/dag_id=ml_pipeline/run_id=scheduled__2023-07-19T22:00:00+00:00/task_id=run_feature_pipeline/attempt=1.log
*** !!!! Please make sure that all your Airflow components (e.g. schedulers, webservers, workers and triggerer) have the same 'secret_key' configured in 'webserver' section and time is synchronized on all your machines (for example with ntpd)
See more at https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#secret-key
*** Could not read served logs: Client error '403 FORBIDDEN' for url 'http://9156d488da30:8793/log/dag_id=ml_pipeline/run_id=scheduled__2023-07-19T22:00:00+00:00/task_id=run_feature_pipeline/attempt=1.log'
For more information check: https://httpstatuses.com/403
```

I missed this error before since it's only there for a few moments and doesn't pop up in the logs when the final error shows.
I am not sure how or why the secret_key or time is not synced - could this be an issue?

**Comment by iusztinpaul** [comment]
I haven't encountered this issue. Thus, it is hard for me to give you step-by-step advice, but here are some docs that I have found after searching the issue on Google:
* https://github.com/bitnami/charts/issues/7167
* https://stackoverflow.com/questions/68907754/access-forbidden-while-accessing-log-in-airflow-with-celeryexecutor

Let me know if this helps.

**Comment by hududed** [comment]
It seems like I have finally found the issue. I completely forgot as I was running each pipeline separately I had setup`ML_PIPELINE_ROOT_DIR` in my bashrc as well. This was the [instruction](https://github.com/iusztinpaul/energy-forecasting#set-up-the-ml_pipeline_root_dir-variable) that I misunderstood leading up to this:

```
gedit ~/.bashrc
export ML_PIPELINE_ROOT_DIR=/path/to/root/directory/repository/energy-forecasting/
```

In other words there was a conflict and the wrong one was spun up. After removing that it seems like this error seems to be solved! Closing this for now
